File: _curso/9_clean_architecture/src/domain/customer/value-object/address.ts
import NotificationError from '../../@shared/notification/notification.error';
import AddressValidatorFactory from '../factory/address.validator.factory';
import Notification from '../../@shared/notification/notification';


export default class Address {
    public notification: Notification;
    _street: string = '';
    _number: number = 0;
    _zip: string = '';
    _city: string = '';

    constructor(street: string, number: number, zip: string, city: string) {
        this._street = street;
        this._number = number;
        this._zip = zip;
        this._city = city;
        this.notification = new Notification();

        this.validate();
        if (this.notification.hasErrors()) {
            throw new NotificationError(this.notification.getErrors());
        }
    }

    get street(): string {
        return this._street;
    }

    get number(): number {
        return this._number;
    }

    get zip(): string {
        return this._zip;
    }

    get city(): string {
        return this._city;
    }

    validate(): void {
        AddressValidatorFactory.create().validate(this);
    }

    toString(): string {
        return `${this._street}, ${this._number}, ${this._zip} ${this._city}`;
    }
}

File: _curso/6_ddd_modelagem_tatica_e_patterns/python/src/domain/customer/value_object/address.py
from dataclasses import dataclass

@dataclass()
class Address:
    _street: str = ''
    _number: int = 0
    _zip: str = ''
    _city: str = ''

    def __post_init__(self):
        self.validate()

    @property
    def street(self) -> str:
        return self._street

    @property
    def number(self) -> int:
        return self._number

    @property
    def zip(self) -> str:
        return self._zip

    @property
    def city(self) -> str:
        return self._city

    def validate(self) -> None:
        if len(self._street) == 0:
            raise Exception("Street is required")

        if self._number == 0:
            raise Exception("Number is required")

        if len(self._zip) == 0:
            raise Exception("Street is required")

        if len(self._city) == 0:
            raise Exception("Street is required")

    def __str__(self) -> str:
        return f"{self._street}, {self._number}, {self._zip} {self._city}"